src/MerchantCenter/PolicyComplianceCheck.php

```diff
@@ -0,0 +1,230 @@
+
+use Automattic\WooCommerce\GoogleListingsAndAds\Infrastructure\Service;
+use Automattic\WooCommerce\GoogleListingsAndAds\Proxies\WC;
+use Automattic\WooCommerce\GoogleListingsAndAds\Google\GoogleHelper;
+
+
+defined( 'ABSPATH' ) || exit;
+
+/**
+ * Class PolicyComplianceCheck
+ *
+ * @package Automattic\WooCommerce\GoogleListingsAndAds\MerchantCenter
+ *
+ * @since x.x.x
+ */
+class PolicyComplianceCheck implements Service {
+	/**
+	 * The WC proxy object.
+	 *
+	 * @var wc
+	 */
+	protected $wc;
+
+	/**
+	 * @var GoogleHelper
+	 */
+	protected $google_helper;
+
+	/**
+	 * BaseController constructor.
+	 *
+	 * @param WC           $wc
+	 * @param GoogleHelper $google_helper
+	 */
+	public function __construct( WC $wc, GoogleHelper $google_helper ) {
+		$this->wc            = $wc;
+		$this->google_helper = $google_helper;
+	}
+
+	/**
+	 * Check if the store website is accessed by all users for the controller.
+	 *
+	 * @return bool
+	 */
+	public function is_accessible(): bool {
+		$all_countries = $this->wc->get_countries();
+		$mc_countries  = $this->google_helper->get_mc_supported_countries();
+
+		foreach ( $mc_countries as $country ) {
+			if ( ! array_key_exists( $country, $all_countries ) ) {
+				return false;
+			}
+		}
+		return true;
+	}
+
+	/**
+	 * Check if the store sample product landing pages lead to a 404 error.
+	 *
+	 * @return bool
+	 */
+	public function has_page_not_found_error(): bool {
+		$ids = wc_get_products(
+			[
+				'return' => 'ids',
+				'limit'  => 1,
+			]
+		);
+		if ( ! empty( $ids ) ) {
+			$url     = get_permalink( $ids[0] );
+			$headers = get_headers( $url );
+			if ( ! empty( $headesr ) && $headers[0] !== 'HTTP/1.1 200 OK' ) {
+				return true;
+			}
+		}
+		return false;
+	}
+
+	/**
+	 * Check if the store sample product landing pages has redirects through 3P domains.
+	 *
+	 * @return bool
+	 */
+	public function has_redirects(): bool {
+		$ids = wc_get_products(
+			[
+				'return' => 'ids',
+				'limit'  => 1,
+			]
+		);
+		if ( ! empty( $ids ) ) {
+			$headers = get_headers( get_permalink( $ids[0] ) );
+			if ( ! empty( $headesr ) && isset( $headers[0] ) ) {
+				if ( $headers[0] === 'HTTP/1.1 302 Found' ) {
+					// this is the URL where it's redirecting
+					return true;
+				}
+			}
+		}
+		return false;
+	}
+
+	/**
+	 * Check if the merchant set the restrictions in robots.txt or not in the store.
+	 *
+	 * @return bool
+	 */
+	public function has_restriction(): bool {
+		return ! $this->robots_allowed( get_site_url() );
+	}
+
+	/**
+	 * Check if the robots.txt has restrictions or not in the store.
+	 *
+	 * @param string $url
+	 * @param bool   $useragent
+	 * @return bool
+	 */
+	private function robots_allowed( $url, $useragent = false ) {
+		// parse url to retrieve host and path
+		$parsed = wp_parse_url( $url );
+
+		$agents = [ preg_quote( '*', null ) ];
+		if ( $useragent ) {
+			$agents[] = preg_quote( $useragent, null );
+		}
+		$agents = implode( '|', $agents );
```

Comment:
We use `$agents` in a regular expression where we use `/` as the delimiter. So we should call `preg_quote` with the delimiter.
```suggestion
		$agents = [ preg_quote( '*', '/' ) ];
		if ( $useragent ) {
			$agents[] = preg_quote( $useragent, '/' );
		}
		$agents = implode( '|', $agents );
```